### PR TITLE
cross-platform diagnostic

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Linuxbrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Linuxbrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Suggested set of changes for Homebrew related to code in `diagnostic.rb`. Instead of `if OS.mac?`,  here we use a function that returns `DYLD` on a Mac and `LD` on Linux.

### CODE REVIEW

**Comment by maxim-belkin** [comment]
I think we could use `HOMEBREW_DYNAMIC_LIBRARY_PREFIX` instead of a function

File: Library/Homebrew/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/diagnostic.rb]: I'm thinking this whole function `check_dyld_vars` should be moved to `os/mac/diagnostic` and `os/linux/diagnostic`, effectively duplicating it. I'm not sure what upstream's preference would be.

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/diagnostic.rb]: I'd suggest the name `generic_check_tmpdir_sticky_bit` rather than `mac_message`.

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/diagnostic.rb]: I suggest
```
create a directory and set the HOMEBREW_TEMP environment variable,
```

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/diagnostic.rb]: I suggest…
```ruby
return if ENV["XDG_DATA_DIRS"].nil? || ENV["XDG_DATA_DIRS"].empty?
return if ENV["XDG_DATA_DIRS"].split(File::PATH_SEPARATOR).include?(HOMEBREW_PREFIX/"share")
```

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/diagnostic.rb]: You can delete these three lines. These methods are called by introspection. They won't be called if they don't exist.

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/diagnostic.rb]: See my above comment about duplicating the whole function `check_dyld_vars`. The new function can be named `check_ld_vars`.

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/diagnostic.rb]: I suggest…
```ruby
return if system f.path
```

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/diagnostic.rb]: Change this line to…
```ruby
export HOMEBREW_TEMP=~/tmp
```

**Review by sjackman — CHANGES_REQUESTED** [review]

**Comment by sjackman** [comment]
Thanks for this PR, Maxim!

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (maxim-belkin) [reply Library/Homebrew/extend/os/linux/diagnostic.rb]: ok

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (maxim-belkin) [reply Library/Homebrew/extend/os/linux/diagnostic.rb]: > set HOMEBREW_TEMP environment variable,

seems ambiguous ... please confirm :)

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (maxim-belkin) [reply Library/Homebrew/extend/os/linux/diagnostic.rb]: ok.

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (maxim-belkin) [reply Library/Homebrew/extend/os/linux/diagnostic.rb]: The idea was to override the default ones with empty ones on Linux where we don't need them...  please confirm.

File: Library/Homebrew/diagnostic.rb
Comment (maxim-belkin) [reply Library/Homebrew/diagnostic.rb]: can you clarify a bit? Are you suggesting renaming `check_dyld_vars` to `check_ld_vars` and using almost identical code in mac and Linux implementations?

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (maxim-belkin) [reply Library/Homebrew/extend/os/linux/diagnostic.rb]: makes sense

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (maxim-belkin) [reply Library/Homebrew/extend/os/linux/diagnostic.rb]: ok

File: Library/Homebrew/diagnostic.rb
Comment (sjackman) [reply Library/Homebrew/diagnostic.rb]: Yep, pretty much. Move `check_dyld_vars` to `os/mac/diagnostic.rb`, and create a new function `check_ld_vars` in `os/linux/diagnostic.rb` that is essentially a duplicate of `check_dyld_vars`.

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (sjackman) [reply Library/Homebrew/extend/os/linux/diagnostic.rb]: You can move these functions from generic to `os/mac/diagnostic.rb`, and `os/linux/diagnostic.rb` does not need any implementation.

**Comment by sjackman** [comment]
Still working on this PR, Maxim?

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (maxim-belkin) [inline Library/Homebrew/extend/os/linux/diagnostic.rb:19]: should I add a call to `message` after `EOS` ?

File: Library/Homebrew/extend/os/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/extend/os/diagnostic.rb]:
```ruby
if OS.mac?
  require "extend/os/mac/diagnostic"
elsif OS.linux?
  require "extend/os/linux/diagnostic"
end
```

File: Library/Homebrew/extend/os/linux/diagnostic.rb
Comment (sjackman) [reply Library/Homebrew/extend/os/linux/diagnostic.rb:19]: It's fine like this.

**Review by sjackman — CHANGES_REQUESTED** [review]
One minor change. Looks good! Thanks, Maxim.

File: Library/Homebrew/extend/os/diagnostic.rb
Comment (maxim-belkin) [reply Library/Homebrew/extend/os/diagnostic.rb]: But the current one takes only two lines instead of 5... Is there any reason to use the classical `if/elsif`?

Comment (sjackman) [reply Library/Homebrew/extend/os/diagnostic.rb]: For consistency with the others. I believe it's Mike's preference.
https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/cleaner.rb

**Comment by sjackman** [comment]
Looks good to me. Can you please squash and rebase?

**Comment by maxim-belkin** [comment]
rebased, squashed, and force-pushed

**Comment by maxim-belkin** [comment]
```
Failures:
  1) Homebrew::Diagnostic::Checks #check_dyld_vars
     Failure/Error: expect(subject.check_dyld_vars).to match("Setting DYLD_INSERT_LIBRARIES")
     NoMethodError:
       undefined method `check_dyld_vars' for #<Homebrew::Diagnostic::Checks:0x00000107bf8238>
     # ./test/diagnostic_spec.rb:165:in `block (2 levels) in <top (required)>'
     # ./test/spec_helper.rb:106:in `block (2 levels) in <top (required)>'
     # ./vendor/bundle/ruby/2.3.0/gems/rspec-retry-0.5.6/lib/rspec/retry.rb:115:in `block in run'
     # ./vendor/bundle/ruby/2.3.0/gems/rspec-retry-0.5.6/lib/rspec/retry.rb:104:in `loop'
     # ./vendor/bundle/ruby/2.3.0/gems/rspec-retry-0.5.6/lib/rspec/retry.rb:104:in `run'
     # ./vendor/bundle/ruby/2.3.0/gems/rspec-retry-0.5.6/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
     # ./vendor/bundle/ruby/2.3.0/gems/rspec-retry-0.5.6/lib/rspec/retry.rb:33:in `block (2 levels) in setup'
     # ./vendor/bundle/ruby/2.3.0/gems/rspec-wait-0.0.9/lib/rspec/wait.rb:46:in `block (2 levels) in <top (required)>'
Finished in 5 minutes 56 seconds (files took 3.33 seconds to load)
1092 examples, 1 failure
Failed examples:
rspec ./test/diagnostic_spec.rb:163 # Homebrew::Diagnostic::Checks #check_dyld_vars
{"uploaded": true, "queued": true, "message": "Coverage reports upload successfully", "id": "da75654d-ae55-4f98-bcfe-10f3f7bf327a", "meta": {"status": 200}, "url": "https://codecov.io/github/Linuxbrew/brew/commit/7e87a653c0b0fd59b350607b426d19c854aca3f3"}
Tests Failed
2268 examples, 1 failure
```

File: Library/Homebrew/os/mac/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/os/mac/diagnostic.rb]: This code should go in the existing file `extend/os/mac/diagnostic.rb` in `Homebrew::Diagnostic::Checks`.

**Review by sjackman — CHANGES_REQUESTED** [review]

File: Library/Homebrew/os/mac/diagnostic.rb
Comment (maxim-belkin) [reply Library/Homebrew/os/mac/diagnostic.rb]: all of it?

Comment (sjackman) [reply Library/Homebrew/os/mac/diagnostic.rb]: Yeah, I believe so. `brew diagnostic` checks live in…
```
diagnostic.rb
extend/os/linux/diagnostic.rb
extend/os/mac/diagnostic.rb
```

**Comment by maxim-belkin** [comment]
Apologies for the delay... updated and force-pushed

File: Library/Homebrew/os/linux/diagnostic.rb
Comment (sjackman) [inline Library/Homebrew/os/linux/diagnostic.rb]: This file is redundant. Move this function to `extend/os/linux/diagnostic.rb`.

**Review by sjackman — CHANGES_REQUESTED** [review]

**Comment by maxim-belkin** [comment]
Failure:

```
#!/bin/bash -eo pipefail
cd /home/linuxbrew/.linuxbrew/Homebrew
git pull --ff-only
if [ -n "$CIRCLE_PR_NUMBER" ]; then
    git fetch origin pull/$CIRCLE_PR_NUMBER/head
fi
git reset --hard $CIRCLE_SHA1
git config --global user.name LinuxbrewTestBot
git config --global user.email testbot@linuxbrew.sh

fatal: Not possible to fast-forward, aborting.
Exited with code 128
```

Can be solved with `git fetch --unshallow` before `git pull`

**Comment by maxim-belkin** [comment]
blocked by #677

**Comment by sjackman** [comment]
Merge PR #678. Should be fixed once the Docker image is updated. https://hub.docker.com/r/linuxbrew/linuxbrew/builds/

**Review by sjackman — APPROVED** [review]

**Comment by sjackman** [comment]
Looks good to me. Thanks, Maxim! I'll merge it after Brew 1.5.14 is tagged. Would you like to submit this same PR upstream to Homebrew/brew?

**Comment by maxim-belkin** [comment]
Yes, I'd like to do that :)

**Comment by sjackman** [comment]
Great. Thanks, Michka!

**Comment by maxim-belkin** [comment]
> Great. Thanks, Michka!

:trollface:

**Comment by sjackman** [comment]
Maxim! Sorry.

**Comment by sjackman** [comment]
Thanks for this work, Maxim! Could you please tag me when you open this PR upstream at Homebrew/brew?

**Comment by maxim-belkin** [comment]
sure

**Pull request merged**
